### PR TITLE
Contextual workflows

### PR DESCRIPTION
**Pull request by mike12345567**
## Description
This introduces the concept of contextual workflows - capable of taking outputs of previous steps/triggers and feeding them as inputs to further steps.

This has been designed to function similar to the data bindings, making use of mustache to describe what the inputs of the various steps should be made up of.

This is built on the concept of each step defining its inputs and outputs in a format as such:

```
schema: {
    inputs: {
      properties: {
        to: {
          type: "string",
          title: "Send To",
        },
        from: {
          type: "string",
          title: "Send From",
        },
        subject: {
          type: "string",
          title: "Email Subject",
        },
        contents: {
          type: "string",
          title: "Email Contents",
        },
      },
      required: ["to", "from", "subject", "contents"],
    },
    outputs: {
      properties: {
        success: {
          type: "boolean",
          description: "Whether the email was sent",
        },
        response: {
          type: "object",
          description: "A response from the email client, this may be an error",
        },
      },
      required: ["success"],
    },
  },
```

This schema should be attached to every trigger, logic or action step in the system and this will be re-used for async blocks in future work.

The front end is capable of displaying these based on the types, titles and descriptions described so in the future adding new steps to a workflow will be very easy - if they are returned in the definitions from the server they will be displayed.

## Screenshots
Here is a video of a relatively simple workflow being built with contextual awareness:
https://www.loom.com/share/8eea5a59bb314ee7871824368d12160b

And a picture of the email that was output:
image

### CODE REVIEW

**Comment by aptkingston** [comment]
💯

File: packages/builder/src/components/workflow/WorkflowBuilder/flowchart/WorkflowBlockTagline.svelte [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    // Add bold tags around inputs
+    let formattedTagline = tagline
+      .replace(/{{/g, "<b>{{")
+      .replace(/}}/, "}}</b>")
+
+    // Extract schema paths for any input bindings
+    const inputPaths = formattedTagline
+      .match(/{{\s*\S+\s*}}/g)
+      .map(x => x.replace(/[{}]/g, "").trim())
+    const schemaPaths = inputPaths.map(x => x.replace(/\./g, ".properties."))
```
Comment (shogunpurple) [inline packages/builder/src/components/workflow/WorkflowBuilder/flowchart/WorkflowBlockTagline.svelte:35]: NAB: try not to use single letter variable names for clarity.

File: packages/server/src/middleware/joi-validator.js [hunk 2]
```diff
@@ -2,9 +2,16 @@ function validate(schema, property) {
   // Return a Koa middleware function
   return (ctx, next) => {
     if (schema) {
-      const { error } = schema.validate(ctx[property])
+      let params =
+        ctx[property] != null
+          ? ctx[property]
+          : ctx.request[property] != null
```
Comment (shogunpurple) [inline packages/server/src/middleware/joi-validator.js:8]: NAB: nested ternaries don’t read well. Might be better to split the separate conditions out into variables.

File: packages/server/src/utilities/accessLevels.js [hunk 3]
```diff
@@ -1,100 +1,36 @@
...
-const adminPermissions = [
+module.exports.READ_MODEL = "read-model"
+module.exports.WRITE_MODEL = "write-model"
+module.exports.READ_VIEW = "read-view"
+module.exports.EXECUTE_WORKFLOW = "execute-workflow"
+module.exports.USER_MANAGEMENT = "user-management"
+module.exports.BUILDER = "builder"
+module.exports.LIST_USERS = "list-users"
+// Access Level IDs
+module.exports.ADMIN_LEVEL_ID = "ADMIN"
```
Comment (shogunpurple) [inline packages/server/src/utilities/accessLevels.js:10]: NAB: You can use an object here to export everything.

```js
 Module.exports = {
   LIST_USERS: “list-users”,
  ..... everything else
}
```

File: packages/server/src/workflows/thread.js [hunk 4]
```diff
@@ -2,18 +2,61 @@ const mustache = require("mustache")
...
+const FILTER_STEP_ID = logic.BUILTIN_DEFINITIONS.FILTER.stepId
+
+function cleanMustache(string) {
+  let charToReplace = {
+    "[": ".",
+    "]": "",
+  }
+  let regex = new RegExp(/{{[^}}]*}}/g)
+  let match
+  while ((match = regex.exec(string)) !== null) {
```
Comment (shogunpurple) [inline packages/server/src/workflows/thread.js:14]: Possibly a good use case for the ES2020 `matchAll`?

File: packages/server/src/workflows/triggers.js [hunk 5]
```diff
@@ -4,7 +4,79 @@ const InMemoryQueue = require("./queue/inMemoryQueue")
 
 let workflowQueue = new InMemoryQueue()
 
-async function queueRelevantWorkflows(event, eventType) {
+const FAKE_STRING = "TEST"
+const FAKE_BOOL = false
```
Comment (shogunpurple) [inline packages/server/src/workflows/triggers.js:8]: What are these fake values for?

**Review by shogunpurple — APPROVED** [review]
Awesome work - you’ve done a great job making a complex piece nice and simple 👍 Just a few questions and optional things.

File: packages/server/src/utilities/accessLevels.js (hunk 3 above)
Comment (mike12345567) [reply packages/server/src/utilities/accessLevels.js:10]: I actually dis-like this method of exporting because it gets into the scenario (sometimes) that access levels was previously in, which is that it can cause circular dependencies.

When the `require` statement is first called the module is configured with its own `module.export` object, which can be mutated to add the properties onto it in the method that I have used.

If you define a new object rather than mutating the existing one you can enter into some pretty nasty circular dependency issues, ones that should be resolved, but there was actually on in the system that I uncovered when I moved some files around where the utility access levels were dependent on the workflow controllers, which unless it was handled this way the access levels module was simply exporting its standard, un-mutated object.

File: packages/server/src/workflows/triggers.js (hunk 5 above)
Comment (mike12345567) [reply packages/server/src/workflows/triggers.js:8]: Basically, something that will need to be solved in the future, but right now we are using the trigger endpoint simply for the builder to be able to test their workflows.

As we are using the trigger endpoint right now to test there is a problem that the `RECORD_SAVED` and `RECORD_DELETED` triggers generate a record as an output when they complete, this can be used as inputs for future stages - when the workflow is triggered from the endpoint the builder does not supply any parameters - it just triggers it, therefore any trigger outputs used in the workflow will be null and the workflow could end up performing differently to how it will perform when it is live.

We thought about whether or not to handle this edge case, I thought with time constraints it would be better to take the work out of the builder and simplify the test experience for the user and simply have the server populate a fake record for the model that was being triggered so that the workflow would perform similarly to what it would in production (just with test data).

Open to input on this, its not the best manner of handling it but I feel something needs to be done for it, otherwise a test workflow button isn't the biggest amount of use, you're better actually going and triggering the workflow (which is also a totally fine way of doing it but I think for some users a simple test workflow button is good).

As a sidenote: In the near future (possibly pre-launch) we may want to allow the user to actually configure a trigger which is the endpoint itself, using the POST body as the output parameters of the trigger which can be used as inputs for future workflow steps. Its a relatively small piece of work to do this, simply creating a new trigger definition for an API endpoint - just need to think about how we would validate this for external apps and how we validate its inputs/outputs.

File: packages/server/src/workflows/thread.js (hunk 4 above)
Comment (mike12345567) [reply packages/server/src/workflows/thread.js:14]: Actually blanked on this one, thanks thats perfect!

File: packages/server/src/middleware/joi-validator.js (hunk 2 above)
Comment (mike12345567) [reply packages/server/src/middleware/joi-validator.js:8]: Yeap you're right, probably for the best!

File: packages/server/src/utilities/accessLevels.js (hunk 3 above)
Comment (shogunpurple) [reply packages/server/src/utilities/accessLevels.js:10]: Ah, makes total sense in that case. Yeah I would eventually like to look at typescript potentially for our server side code or even just run babel over it so we can get some nice bits like ESM.

Comment (mike12345567) [reply packages/server/src/utilities/accessLevels.js:10]: Yeap running Babel over the top is a good solution, means you can use stuff like [babel root plugins](https://github.com/entwicklerstube/babel-plugin-root-import) as well which can really help in multi-layered projects!

File: packages/server/src/workflows/triggers.js (hunk 5 above)
Comment (shogunpurple) [reply packages/server/src/workflows/triggers.js:8]: Ah, see what you mean. That's perfect - I'd imagine we will complete most of that work you mentioned as part of https://github.com/Budibase/budibase/issues/495. Anything that pertains to the external execution of workflows via HTTP would make sense to be solved there.

This works great for testing workflows - I actually think this is a good way to identify the test record when it's created so you can just delete it again.
